### ISSUE TITLE
Switching PhantomJS to sharp has broken SVG scaling

### ISSUE THREAD

**Issue by ahukkanen** [itgalaxy/favicons#264]
## Description of the issue

After switching PhantomJS to sharp, SVG scaling has been broken. In case a lower resolution SVG is provided as the source image than the largest generated icon, the generated icons are blurred as the lower resolution image is scaled up.

Here is an example of the result with a sample of a high resolution generated icon:
![scaling-issue](https://user-images.githubusercontent.com/864340/57950869-24b7fd80-78f1-11e9-98dd-c53ef3418c1b.png)

The icons have been generated with an SVG with the following source:

```svg
<?xml version="1.0" encoding="UTF-8"?>
<svg width="64" height="64" clip-rule="evenodd" image-rendering="optimizeQuality" shape-rendering="geometricPrecision" text-rendering="geometricPrecision" version="1.1" viewBox="0 0 64 64" xmlns="http://www.w3.org/2000/svg"><rect width="64" height="64" fill="#fff" style="paint-order:stroke fill markers"/><path d="m8.9671 31.428 5.4166 5.5052a20.452 20.452 0 0 1 1.8508-1.8969c3.9921-3.5988 9.522-5.8252 15.644-5.8252v0.0153h0.0012v-0.0153c6.1217 0 11.653 2.2264 15.645 5.8252 0.73699 0.6662 1.4221 1.3772 2.047 2.1308l5.4638-5.4627a28.333 28.333 0 0 0-2.3575-2.393c-5.3351-4.8083-12.691-7.7847-20.798-7.7847v-0.0142h-0.0012v0.0142c-8.1083 0-15.463 2.9764-20.797 7.7847a27.925 27.925 0 0 0-2.1142 2.1166zm23.147 2.719v0.0153c5.5512 1e-3 10.59 2.0386 14.25 5.3327v1e-3l-4.4647 4.4658-1.0299 0.9626c-2.2866-1.9099-5.361-3.0768-8.7555-3.0768v0.0154-0.0154c-3.5658 0-6.7807 1.2886-9.0969 3.3721-0.0578 0.052-0.1145 0.1051-0.1724 0.1583l-5.2076-5.6753c0.0745-0.0697 0.15-0.1394 0.2268-0.2079 3.6591-3.2941 8.7-5.3315 14.25-5.3327v-0.0154zm0.144 12.045c2.2206 0 4.2403 0.8586 5.7474 2.2606l-6.4937 6.6875-5.5158-6.1655c1.5449-1.709 3.7784-2.7827 6.2622-2.7827zm-27.047-19.252a38.128 38.128 0 0 1 1.0063-0.9437c6.4689-5.83 15.421-9.4371 25.323-9.4371v0.0142-0.0142c9.9013 0 18.854 3.6071 25.324 9.4371a35.059 35.06 0 0 1 1.689 1.6335l5.4485-5.445a42.135 42.135 0 0 0-1.9843-1.9134c-7.8143-7.0406-18.59-11.397-30.477-11.397v-0.0153 0.0153c-11.888 0-22.663 4.356-30.475 11.397a42.522 42.522 0 0 0-1.0654 0.9934l5.2111 5.6752z" fill="#295682" stroke-width=".1"/></svg>
```

Save that to a file named `testicon.svg` and run the favicon generation.

This icon is a modified version of the following SVG (that can be used as well to generate the same issue):
https://uxwing.com/wifi-icon/

## Further information

This issue originates from this pull request:
https://github.com/itgalaxy/favicons/pull/254

And from this commit:
https://github.com/itgalaxy/favicons/commit/928524c1abd57f2fb2903ab69f379f1123eb8c6a

The versions before that commit work properly regarding the explained issue.

## Example code to replicate the issue

package.json (broken):

```json
{
  "name": "favicons-bug",
  "version": "0.0.1",
  "description": "",
  "main": "favicons.js",
  "scripts": {
    "favicons": "node favicons.js"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "favicons": "github:itgalaxy/favicons#928524c1abd57f2fb2903ab69f379f1123eb8c6a",
    "fs-extra": "^8.0.1"
  }
}
```

package.json (working):

```json
{
  "name": "favicons-bug",
  "version": "0.0.1",
  "description": "",
  "main": "favicons.js",
  "scripts": {
    "favicons": "node favicons.js"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "favicons": "github:itgalaxy/favicons#3137f1f5a4c0c64f7368203be3dcbddead6a27ef",
    "fs-extra": "^8.0.1"
  }
}
```

favicons.js:

```js
const favicons = require('favicons');
const fs = require("fs-extra");

const source = `${__dirname}/images/testicon.svg`;
const targetPath = `${__dirname}/favicons`;

const callback = (error, response) => {
  if (error) {
    console.log(error.message);
    return;
  }

  response.images.forEach((image) => {
    // Write the actual image files
    const path = `${targetPath}/${image.name}`;
    const stream = fs.createWriteStream(path);
    stream.once("open", () => {
      stream.write(image.contents);
      stream.end();
    });
  });
};

// Execute the process in order
fs.remove(targetPath)
  .then(() => fs.mkdir(targetPath))
  .then(() => favicons(source, {}, callback))
;
```

## Workaround

A workaround exists for this problem which is to define the SVG in higher resolution or equal to the largest generated icon.

**Comment by ahukkanen** [comment]
Possibly @kghost would have further information regarding this issue as the author of #254?

**Comment by ahukkanen** [comment]
It seems sharp is actually rendering the SVG image with 72DPI in its originally defined width and height:
https://github.com/lovell/sharp/issues/729

I tried the suggestion to provide the calculated resizes density option to the constructor but it does not seem to provide any better result:
https://github.com/lovell/sharp/issues/729#issuecomment-284708688

My best guess how to fix this would be to modify the original SVG's width and height before passing it to sharp here:
https://github.com/itgalaxy/favicons/blob/7d1ff2f5af00ef7b15b1dbe7d7ba438a664c9d59/src/helpers.js#L217

If that was done, I believe sharp would render the SVG in its desired size. I will try and experiment a bit with this.

**Comment by ahukkanen** [comment]
My guess (https://github.com/itgalaxy/favicons/issues/264#issuecomment-493680063) was correct and seems to fix the issue.

There is now a PR available to implement that fix: #265

**Comment by ahukkanen** [comment]
For further information, I also dug a bit into `svg2png` and it seems that it is actually doing something similar:
https://github.com/domenic/svg2png/blob/5997ae9452796f480c13506c32cb04d1dc38369e/lib/converter.js#L84-L105

So, they are setting the SVG dimensions for the rendering directly to the SVG element, so that PhantomJS renders it in the expected size. Of course, from their perspective it is sort of required as they are not actually doing any image processing. So if they want to render an SVG in the targeted size, they need to pass it with that size to the "renderer" (PhantomJS).

The difference with sharp is that it first renders the SVG using `librsvg` (or actually with `libvips` which is using `librsvg` for SVGs) and **then** scales the rendered image up in case the target size is higher than the rendered size. It sounds it could be also something that could be directly fixed to sharp but for the time being, #265 can be used as a workaround.

I added a comment about this in a related (closed) issue in sharp's repo:
https://github.com/lovell/sharp/issues/1421#issuecomment-493707752

It might also be from sharp's perspective this issue is out of their scope.

The good thing about the workaround is that it does not add any extra dependencies to the library. The `xml2js` is already a dependency through `jimp -> load-bmfont -> parse-bmfont-xml -> xml2js` and `to-ico -> resize-img -> jimp -> ...`.
